[CCM - service controller] addressing left over comments from #109706

alexanderConstantinescu opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind bug 

#### What this PR does / why we need it:

This is a follow up from #109706 which extends and reduces the node re-sync predicates for ETP=local services to avoid re-syncs and LB re-configuration for taint and schedulability changes. The conclusion from the referenced PR was: we should avoid re-syncing LBs as much as possible (because that's costly for what concerns service SLA's/downtime) and have the LB rely on the defined HC that goes with ETP=local services. That PR already removed the predicate for the node's readiness state, this PR extends that logic to the aforementioned taint conditions.   

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->

https://github.com/kubernetes/kubernetes/issues/111824

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Expands the partial fix for https://github.com/kubernetes/kubernetes/issues/111539 which was already started in https://github.com/kubernetes/kubernetes/pull/109706 Specifically, we will now reduce the amount of syncs for ETP=local services even further in the CCM and avoid re-configuring LBs to an even greater extent.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

/sig network
/sig cloud-provider
/assign @thockin

alexanderConstantinescu committed:
Address follow up comments from #109706

Specifically:

- https://github.com/kubernetes/kubernetes/pull/109706#discussion_r936113657
- https://github.com/kubernetes/kubernetes/pull/109706#discussion_r936092033

thockin commented on staging/src/k8s.io/cloud-provider/controllers/service/controller.go:
```diff
@@ -987,8 +987,6 @@ var (
 	}
 	etpLocalNodePredicates []NodeConditionPredicate = []NodeConditionPredicate{
 		nodeIncludedPredicate,
-		nodeSchedulablePredicate,
```
So if I understand, this will immeditately cause a sync when CCM restarts if:

* etp=Local AND
* node is not schedulable or is tainted

Right?  What I mean is that a node is currently NOT in the LB set because of the old predicates, but under the new predicates we would add it to the LB set.  This would cause it start accepting traffic (because the HC changes are not in yet) and then stop serving traffic again when the node was terminated.

Is that a correct assessment?

For eTP=Local that seems like a small risk, but when we want to do this for all services, more important.  I guess that, by then, we'll have the HC changes in place.

alexanderConstantinescu replied:
> This would cause it start accepting traffic (because the HC changes are not in yet)

Yes but the node would also have to be hosting the running pod - otherwise it wouldn't get any traffic sent its way because of the ETP=local HC. Worst case scenario: we re-configure the LB with the not schedulable node, traffic is sent to it because it hosts the currently running pod, but pod is very shortly after stopped. This is not really a problem at all, right? This _might_ be a problem for the readiness predicate though, because maybe the node is not ready because networking is bonkers, but at the same time: it's able to answer to the LB HC, so it's not really that bonkers either 🤷 

> but when we want to do this for all services, more important.

Yes, because we wouldn't be protected by the ETP=local HC. So we'll add a potential "unsafe" (is a "not schedulable node" really "unsafe"?) node back into the LB set, which - worst case scenario - might not be able to forward traffic anywhere...but again, just how bad have things _really_ gone networking wise if the node is still able to send a response to the LB HC?

thockin reviewed (commented):
For the record: why I think this is "safe" in general (even for non eTP=Local Services)

Given:
* Nodes N1, N2, N3 are ready, schedulable, untainted, and un-excluded
* Service S has LB sending to {N1, N2, N3}
 
Today, what happens is:
* If N1 becomes unschedulable (or any other of the above) or is deleted in k8s:
   * Service S LB set is chaged to {N2, N3} (N1 is removed)
   * Any open connections from LB -> N1 will break abruptly without warning 
   * No new connections will be made
* If N1 suddently powers off or is network partitioned:
   * Any open connections from LB -> N1 will break abruptly without warning 
   * New connections will fail
   * The LB->Node healthcheck will eventually fail and no new connections will be made

After this change:
* If N1 becomes unschedulable (or any other of the above):
   * Service S LB set is untouched
   * Open connections from LB -> N1 will continue
   * New connections will be allowed
* If N1 is deleted in k8s:
   * Service S LB set is chaged to {N2, N3} (N1 is removed)
   *  Any open connections from LB -> N1 will break abruptly without warning 
   * No new connections will be made
* If N1 suddently powers off or is network partitioned:
   * Any open connections from LB -> N1 will break abruptly without warning 
   * New connections will fail
   * The LB->Node healthcheck will eventually fail and no new connections will be made

So the new result is no worse than before.  The abrupt termination of connections just happens at a different point in time.

Once we get the kube-proxy healthcheck URL to return "not-ready" when unschedulable, tainted, etc. Then it will be even better:
* If N1 becomes unschedulable (or any other of the above):
   * Service S LB set is untouched
   * Open connections from LB -> N1 will continue
   * New connections will be allowed until the HC fires, at which point:
   * Open connections from LB -> N1 will continue
   * No new connections will be made
* If N1 is deleted in k8s:
   * Service S LB set is chaged to {N2, N3} (N1 is removed)
   *  Any open connections from LB -> N1 will break abruptly without warning 
   * No new connections will be made
* If N1 suddently powers off or is network partitioned:
   * Any open connections from LB -> N1 will break abruptly without warning 
   * New connections will fail
   * The LB->Node healthcheck will eventually fail and no new connections will be made

So...

/lgtm
/approve

But I do not think we should push for this in 1.25 - it's the low-value part of the proposal and the code-freeze is on.  Release team has enough to deal with, without us making trouble.

thockin commented:
/ok-to-test
/hold for #109706

thockin commented:
Agree.  I am just worried about the transition and avoiding a full resync
of all nodes in all services, if possible.

On Thu, Aug 4, 2022 at 12:31 AM Alexander Constantinescu <
***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In staging/src/k8s.io/cloud-provider/controllers/service/controller.go
> <https://github.com/kubernetes/kubernetes/pull/111658#discussion_r937443439>
> :
>
> > @@ -987,8 +987,6 @@ var (
>
>  	}
>
>  	etpLocalNodePredicates []NodeConditionPredicate = []NodeConditionPredicate{
>
>  		nodeIncludedPredicate,
>
> -		nodeSchedulablePredicate,
>
>
> This would cause it start accepting traffic (because the HC changes are
> not in yet)
>
> Yes but the node would also have to be hosting the running pod - otherwise
> it wouldn't get any traffic sent its way because of the ETP=local HC. Worst
> case scenario: we re-configure the LB with the not schedulable node,
> traffic is sent to it because it hosts the currently running pod, but pod
> is very shortly after stopped. This is not really a problem at all, right?
> This *might* be a problem for the readiness predicate though, because
> maybe the node is not ready because networking is bonkers, but at the same
> time: it's able to answer to the LB HC, so it's not really that bonkers
> either 🤷
>
> but when we want to do this for all services, more important.
>
> Yes, because we wouldn't be protected by the ETP=local HC. So we'll add a
> potential "unsafe" (is a "not schedulable node" really "unsafe"?) node back
> into the LB set, which - worst case scenario - might not be able to forward
> traffic anywhere...but again, just how bad have things *really* gone
> networking wise if the node is still able to send a response to the LB HC?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/111658#discussion_r937443439>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKWAVA2GFH3SQMI7DN65GTVXNWVTANCNFSM55NZTAUQ>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

alexanderConstantinescu commented:
@thockin : I think I understand the `nodeUnTaintedPredicate` much better now. We should not be removing it. 

In a world without this predicate we would have:

1. Cluster auto-scaler starts deleting node and associated VM instance
2. Cloud finishes GC'ing VM instance
3. Cluster auto-scaler triggers Node delete.
4. CCM gets triggered by the Node deletion event - issues a call to delete node from the LB set 
5. CCM / Kube cloud-provider code / real cloud provider - starts erroring because it can't find the associated instance, because it got deleted in step 2 and can't remove it from the LB set. 

I am not sure if all clouds GC all associations to the VM instance when it gets deleted. If that is the case: we would need to have the cloud-provider code be more robust and compare the current state of the LB set and skip the update call to the cloud, should the node not be in the set anymore. That would remove this predicate and allow us to reconcile a node deletion without errors.    

I can add this as a doc bloc to that predicate. But we should not remove it, for now.

alexanderConstantinescu commented:
/retitle [CCM - service controller] addressing left over comments from #109706

thockin commented:
> I think I understand the nodeUnTaintedPredicate much better now. We should not be removing it.

I don't know the exact end-of-life sequence for autoscaler, but it seems weird to delete the VM before deleting the k8s Node.  If that is true, then you're right - we have to remove it from the LB set before the VM can be deleted (otherwise it is "in use" on at least some clouds).

But the goal of NOT removing it is to allow the LB HC to fail the node and "drain" traffic (allow existing but no new connections).  

IOW: I think the sequence _should_ be:

1) taint node
  1a) CCM does nothing
  1b) kube-proxy starts failing HC
  1c) LBs stop sending new requests, but leave existing ones
2) mark node unschedulable
3) drain any running pods
4) delete Node
  4a) CCM removes node from LB sets
  4b) LBs stop sending new connections, break existing ones
5) delete underlying VM

Is that wrong?  Maybe auto-scaling SIG can clarify?

@gjtempleton 
@mwielgus

alexanderConstantinescu commented:
> but it seems weird to delete the VM before deleting the k8s Node.

Having skimmed through the autoscaler code: I confirm that to be the case. It taints the node with that key, then proceeds to delete the VM from the node group, see: [taint + async delete](https://github.com/kubernetes/autoscaler/blob/5cf8cbdf52fab5a731127049f8ba992af7ad1e3d/cluster-autoscaler/core/scaledown/actuation/actuator.go#L139) / [taint + sync delete](https://github.com/kubernetes/autoscaler/blob/5cf8cbdf52fab5a731127049f8ba992af7ad1e3d/cluster-autoscaler/core/scaledown/actuation/actuator.go#L172), it doesn't explicitly delete the k8s node though, so how that happens remains unclear to me. 

> IOW: I think the sequence should be:
> 
> taint node
> 1a) CCM does nothing
> 1b) kube-proxy starts failing HC
> 1c) LBs stop sending new requests, but leave existing ones
> mark node unschedulable
> drain any running pods
> delete Node (with or without finalizer)
> 4a) CCM removes node from LB sets
> 4b) LBs stop sending new connections, break existing ones
> delete underlying VM
 
If we manage to do what we want, we could also just do:

taint node
1a) CCM does nothing
1b) kube-proxy does nothing
mark node unschedulable
drain any running pods
3a) kube-proxy starts failing HC - because it's not running on the node anymore 
3b) LBs stop sending new requests, but leave existing ones
delete Node (with or without finalizer)
4a) CCM removes node from LB sets
4b) LBs stop sending new connections, break existing ones
delete underlying VM

That would avoid having kube-proxy react to the taints on the node object. But it doesn't seem to me like we will be able to do 1a)

alexanderConstantinescu commented:
OK, I just found time to focus back on this and just revisited the PR. 

What [I mentioned above is](https://github.com/kubernetes/kubernetes/pull/111658#issuecomment-1211014732) completely wrong. I got so caught up in the "Node is added to the LB set / Node is removed from the LB set" that I completely overlooked the fact that we can only "overwrite the full LB set". Worrying about "deleting a node from the LB set that might not have a VM anymore and causing the LB sync to fail" can be proven completely useless by just looking at the LoadBalancer API defined in `cloud-provider`:

```go
// LoadBalancer is an abstract, pluggable interface for load balancers.
//
// Cloud provider may chose to implement the logic for
// constructing/destroying specific kinds of load balancers in a
// controller separate from the ServiceController.  If this is the case,
// then {Ensure,Update}LoadBalancer must return the ImplementedElsewhere error.
// For the given LB service, the GetLoadBalancer must return "exists=True" if
// there exists a LoadBalancer instance created by ServiceController.
// In all other cases, GetLoadBalancer must return a NotFound error.
// EnsureLoadBalancerDeleted must not return ImplementedElsewhere to ensure
// proper teardown of resources that were allocated by the ServiceController.
// This can happen if a user changes the type of LB via an update to the resource
// or when migrating from ServiceController to alternate implementation.
// The finalizer on the service will be added and removed by ServiceController
// irrespective of the ImplementedElsewhere error. Additional finalizers for
// LB services must be managed in the alternate implementation.
type LoadBalancer interface {
	// TODO: Break this up into different interfaces (LB, etc) when we have more than one type of service
	// GetLoadBalancer returns whether the specified load balancer exists, and
	// if so, what its status is.
	// Implementations must treat the *v1.Service parameter as read-only and not modify it.
	// Parameter 'clusterName' is the name of the cluster as presented to kube-controller-manager
	GetLoadBalancer(ctx context.Context, clusterName string, service *v1.Service) (status *v1.LoadBalancerStatus, exists bool, err error)
	// GetLoadBalancerName returns the name of the load balancer. Implementations must treat the
	// *v1.Service parameter as read-only and not modify it.
	GetLoadBalancerName(ctx context.Context, clusterName string, service *v1.Service) string
	// EnsureLoadBalancer creates a new load balancer 'name', or updates the existing one. Returns the status of the balancer
	// Implementations must treat the *v1.Service and *v1.Node
	// parameters as read-only and not modify them.
	// Parameter 'clusterName' is the name of the cluster as presented to kube-controller-manager
	EnsureLoadBalancer(ctx context.Context, clusterName string, service *v1.Service, nodes []*v1.Node) (*v1.LoadBalancerStatus, error)
	// UpdateLoadBalancer updates hosts under the specified load balancer.
	// Implementations must treat the *v1.Service and *v1.Node
	// parameters as read-only and not modify them.
	// Parameter 'clusterName' is the name of the cluster as presented to kube-controller-manager
	UpdateLoadBalancer(ctx context.Context, clusterName string, service *v1.Service, nodes []*v1.Node) error
	// EnsureLoadBalancerDeleted deletes the specified load balancer if it
	// exists, returning nil if the load balancer specified either didn't exist or
	// was successfully deleted.
	// This construction is useful because many cloud providers' load balancers
	// have multiple underlying components, meaning a Get could say that the LB
	// doesn't exist even if some part of it is still laying around.
	// Implementations must treat the *v1.Service parameter as read-only and not modify it.
	// Parameter 'clusterName' is the name of the cluster as presented to kube-controller-manager
	EnsureLoadBalancerDeleted(ctx context.Context, clusterName string, service *v1.Service) error
}
```

When a "Node is removed from the LB set" the only calls possibly by the KCCM are: `EnsureLoadBalancer` or `UpdateLoadBalancer`, both of which take `nodes` as an argument consisting of all the __existing__ nodes the caller intends to configure. The KCCM/cloud-provider won't even lookup the VM associated with the deleted node.   

I have hence again removed the `nodeUnTaintedPredicate` was we originally discussed and agreed upon.

thockin commented:
I'm trying to get an authoritative answer in https://github.com/kubernetes/autoscaler/issues/5201 - my reading is that it sets the taint and `unschedulable` at the same time. 

If we respect the taint predicate, then cluster-autoscaler scale-down events will not have LB draining and will immediately be removed from LBs.

If we don't respect the taint predicate, then cluster-autoscaler scale-down events will have LB draining, but (I think) the VM deletion will fail becuase it is still in the LB set.

thockin commented on staging/src/k8s.io/cloud-provider/controllers/service/controller.go:
```diff
@@ -989,7 +989,6 @@ var (
 	}
 	etpLocalNodePredicates []NodeConditionPredicate = []NodeConditionPredicate{
 		nodeIncludedPredicate,
-		nodeUnTaintedPredicate,
```
based on my read of cluster-autoscaler, I think we need to keep this.  Let's let that discussion play out.

thockin commented:
Re-posting revised drawing, ICYMI on the bug.  Undure of the taint part, but I do think we should handle terminationGracePeriod (as a new predicate?)

![Node_LB_states_v2](https://user-images.githubusercontent.com/5595220/191381368-6db6831a-2a90-4991-8f50-645c03afd5b7.png)

thockin commented:
Discussion of the taint in https://github.com/kubernetes/autoscaler/issues/5201

dims commented:
If you still need this PR then please rebase, if not, please close the PR

alexanderConstantinescu commented:
OK, I've rebased and removed the second commit. Though we are lacking a conclusion on https://github.com/kubernetes/autoscaler/issues/5201, it does seem that we should not be removing that taint from the service controller sync loop. So for now, this PR just addresses the left over comments 

PTAL: @thockin

thockin commented:
Updated thread on autoscaler issue

thockin commented:
Now that a lot of the other changes are in, this PR is reduced to some cleanup

/lgtm
/approve

@alexanderConstantinescu any objections?  If not, remove hold?

aojea commented:
failure seems to be https://github.com/kubernetes/kubernetes/issues/107414

/retest
